scrapy/pipelines/__init__.py

```diff
@@ -21,7 +21,8 @@ def _get_mwlist_from_settings(cls, settings):
                 category=ScrapyDeprecationWarning, stacklevel=1)
             # convert old ITEM_PIPELINE list to a dict with order 500
             item_pipelines = dict(zip(item_pipelines, range(500, 500+len(item_pipelines))))
-        return build_component_list(settings['ITEM_PIPELINES_BASE'], item_pipelines)
+            settings.set('ITEM_PIPELINES', item_pipelines)
```

Comment:
There are a couple of issues here:
* There's no guarantee that `settings.set('ITEM_PIPELINES', item_pipelines)` overrides the stored ITEM_PIPELINES, that `settings.set()` call should be made with a priority higher than the stored one for that key.
* We didn't modify the settings before, and not too long ago the settings were frozen at this point, this would have thrown an error. A solution for it could be to make a copy and set ITEM_PIPELINES there.
* This is incompatible with the goal of allowing addons to modify settings, addons won't be able to add pipelines easily if ITEM_PIPELINES is a list.

My vote is to remove all this deprecation support for lists, it was introduced way too long ago (0.20 release: fc388f46).